api/layer/multipart_upload.go

```diff
@@ -0,0 +1,563 @@
+		meta, err := n.objectHead(ctx, p.Bkt.CID, id)
+		if err != nil {
+			n.log.Warn("couldn't head object",
+				zap.Stringer("object id", id),
+				zap.Stringer("bucket id", p.Bkt.CID),
+				zap.Error(err))
+			continue
+		}
+		info := uploadInfoFromMeta(meta, p.Prefix, p.Delimiter)
+		if info != nil {
+			uploads = append(uploads, info)
+		}
+	}
+
+	sort.Slice(uploads, func(i, j int) bool {
+		if uploads[i].Key == uploads[j].Key {
+			return uploads[i].UploadID < uploads[j].UploadID
+		}
+		return uploads[i].Key < uploads[j].Key
+	})
+
+	if p.KeyMarker != "" {
+		if p.UploadIDMarker != "" {
+			uploads = trimAfterUploadIDAndKey(p.KeyMarker, p.UploadIDMarker, uploads)
+		} else {
+			uploads = trimAfterUploadKey(p.KeyMarker, uploads)
+		}
+	}
+
+	if len(uploads) > p.MaxUploads {
+		result.IsTruncated = true
+		uploads = uploads[:p.MaxUploads]
+		result.NextUploadIDMarker = uploads[len(uploads)-1].UploadID
+		result.NextKeyMarker = uploads[len(uploads)-1].Key
+	}
+
+	uniqDirs := make(map[string]struct{})
+
+	for _, ov := range uploads {
+		if ov.IsDir {
+			if _, ok := uniqDirs[ov.Key]; !ok {
+				uniqDirs[ov.Key] = struct{}{}
+				result.Prefixes = append(result.Prefixes, ov.Key)
+			}
+		} else {
+			result.Uploads = append(result.Uploads, ov)
+		}
+	}
+
+	return &result, nil
+}
+
+func (n *layer) AbortMultipartUpload(ctx context.Context, p *UploadInfoParams) error {
+	objects, err := n.getUploadParts(ctx, p)
+	if err != nil {
+		return err
+	}
+
+	for _, info := range objects {
+		err := n.objectDelete(ctx, info.CID, info.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (n *layer) ListParts(ctx context.Context, p *ListPartsParams) (*ListPartsInfo, error) {
+	var res ListPartsInfo
+	objs, err := n.getUploadParts(ctx, p.Info)
+	if err != nil {
+		return nil, err
+	}
+
+	res.Owner = objs[0].Owner
+
+	parts := make([]*Part, 0, len(objs))
+
+	for num, objInfo := range objs {
+		if num == 0 {
+			continue
+		}
+		parts = append(parts, &Part{
+			ETag:         objInfo.HashSum,
+			LastModified: objInfo.Created.Format(time.RFC3339),
+			PartNumber:   num,
+			Size:         objInfo.Size,
+		})
+	}
+
+	sort.Slice(parts, func(i, j int) bool {
+		return parts[i].PartNumber < parts[j].PartNumber
+	})
+
+	if p.PartNumberMarker != 0 {
+		for i, part := range parts {
+			if part.PartNumber > p.PartNumberMarker {
+				parts = parts[i:]
+				break
+			}
+		}
+	}
+
+	if len(parts) > p.MaxParts {
+		res.IsTruncated = true
+		res.NextPartNumberMarker = parts[p.MaxParts-1].PartNumber
+		parts = parts[:p.MaxParts]
+	}
+
+	res.Parts = parts
+
+	return &res, nil
+}
+
+func (n *layer) GetUploadInitInfo(ctx context.Context, p *UploadInfoParams) (*data.ObjectInfo, error) {
+	ids, err := n.objectSearchByName(ctx, p.Bkt.CID, createUploadPartName(p.UploadID, p.Key, 0))
+	if err != nil {
+		return nil, err
+	}
+	if len(ids) == 0 {
+		return nil, errors.GetAPIError(errors.ErrNoSuchUpload)
+	}
+
+	meta, err := n.objectHead(ctx, p.Bkt.CID, ids[0])
```

Comment:
I'm not sure we can return the first found object. The error seems to be more appropriate if we found more than one initial part.

Reply:
The case when we found two objects with the same upload id and name is odd, but okay